src/Ledger/Epoch.agda

```diff
@@ -7,98 +7,108 @@ open import Ledger.Prelude hiding (compare; Rel)
 
-  instance
-    preoSlot : HasPreorder Slot _≡_
-    preoSlot = record { _≤_ = _≤ˢ_ ; isPreorder = ≤ˢ-isPreorder }
+  ≤ˢ-reflexive : Reflexive _≤ˢ_
+  ≤ˢ-reflexive = IsPreorder.reflexive ≤ˢ-isPreorder refl
 
-    poSlot : HasPartialOrder Slot _≡_
-    poSlot = record { hasPreorder = preoSlot ; antisym = ≤ˢ-isAntisymmetric }
+  ≤ˢ-transitive : Transitive _≤ˢ_
+  ≤ˢ-transitive = IsPreorder.trans ≤-isPreorder
 
-  Dec≡⋀TotAntisym≤⇒Dec≤ : Decidable _≡_ → Antisymmetric _≡_ _≤ˢ_ → Total _≤ˢ_ → Decidable _≤ˢ_
-  Dec≡⋀TotAntisym≤⇒Dec≤ dec≡ antisym≤ tot≤ x y with dec≡ x y | tot≤ x y
-  ... | yes refl | _ = true because ofʸ (IsPartialOrder.reflexive ≤ˢ-isPartialOrder refl)
-  ... | no ¬p | inj₁ x≤y = true because ofʸ x≤y
-  ... | no ¬p | inj₂ y≤x = false because (ofⁿ (λ x≤y → ¬p (antisym≤ x≤y y≤x)))
+  ≤ˢ-isTotalOrder : IsTotalOrder _≡_ _≤ˢ_
+  ≤ˢ-isTotalOrder = <-STO⇒≤-isTotalOrder Slot _≡_ _<ˢ_ Slot-STO
 
+  ≤ˢ-total : Total _≤ˢ_
+  ≤ˢ-total = IsTotalOrder.total ≤ˢ-isTotalOrder
 
   _≤ˢ?_ : (s s' : Slot) → Dec (s ≤ˢ s')
-  _≤ˢ?_ = Dec≡⋀TotAntisym≤⇒Dec≤ _≡ˢ?_ ≤ˢ-isAntisymmetric ≤ˢ-isTotal
+  _≤ˢ?_ = _≤?_
 
   ≤ˢ-isDecTotalOrder : IsDecTotalOrder _≡_ _≤ˢ_
-  ≤ˢ-isDecTotalOrder = record { isTotalOrder = Slot-TO ; _≟_ = _≡ˢ?_ ; _≤?_ = _≤ˢ?_ }
+  ≤ˢ-isDecTotalOrder = record { isTotalOrder = ≤ˢ-isTotalOrder ; _≟_ = _≡ˢ?_ ; _≤?_ = _≤ˢ?_ }
 
   instance
-    Dec-≤ˢ : ∀ {n m} → Dec (n ≤ˢ m)
-    Dec-≤ˢ = Decidable²⇒Dec _≤ˢ?_
-
-    decpoSlot : HasDecPartialOrder Slot _≡_
-    decpoSlot = record { hasPartialOrder = poSlot ; _≤?_ = _≤ˢ?_ }
+    Dec-<ˢ : ∀ {n m : Slot} → Dec (n <ˢ m)
+    Dec-<ˢ = Decidable²⇒Dec (IsStrictTotalOrder._<?_ Slot-STO)
 
   _≤ᵉ_ : Epoch → Epoch → Set
-  e ≤ᵉ e' = firstSlot e ≤ˢ firstSlot e'
-
-  ≤ᵉ-isReflexive : Reflexive _≤ᵉ_
-  ≤ᵉ-isReflexive = IsPreorder.reflexive ≤ˢ-isPreorder refl
+  e ≤ᵉ e' = firstSlot e ≤ firstSlot e'
 
-  _≤ᵉ?_ : ∀ e e' → Dec (e ≤ᵉ e')
-  e ≤ᵉ? e' = firstSlot e ≤ˢ? firstSlot e'
+  ≤ᵉ-reflexive : Reflexive _≤ᵉ_
+  ≤ᵉ-reflexive = ≤ˢ-reflexive
 
   ≤ᵉ-isPreorder : IsPreorder _≡_ _≤ᵉ_
   ≤ᵉ-isPreorder .IsPreorder.isEquivalence     = Ledger.Prelude.isEquivalence
-  ≤ᵉ-isPreorder .IsPreorder.reflexive refl    = ≤ᵉ-isReflexive
-  ≤ᵉ-isPreorder .IsPreorder.trans ij jk       = ≤ˢ-isTransitive ij jk
-
-  _ = (∀ {n m} → Dec (n <ˢ m)) ∋ it
-  _ = (∀ {n m} → Dec (n ≤ˢ m)) ∋ it
-  _ = (∀ {n m} → Dec (n ≤ᵉ m)) ∋ it
+  ≤ᵉ-isPreorder .IsPreorder.reflexive refl    = ≤ᵉ-reflexive
+  ≤ᵉ-isPreorder .IsPreorder.trans ij jk       = ≤ˢ-transitive ij jk
 
   instance
     preoEpoch : HasPreorder Epoch _≡_
-    preoEpoch = record { _≤_ = _≤ᵉ_ ; isPreorder = ≤ᵉ-isPreorder }
+    preoEpoch = hasPreorderFromNonStrict Epoch _≡_ _≤ᵉ_ ≤ᵉ-isPreorder _≡ᵉ?_
+
+  _<ᵉ_ : Rel Epoch _
+  _<ᵉ_ = _<_
+
+  _ = ({s s' : Slot} → Dec (s < s')) ∋ it
+  _ = ({s s' : Slot} → Dec (s ≤ s')) ∋ it
+  _ = ({e e' : Epoch} → Dec (e < e')) ∋ it
+  _ = ({e e' : Epoch} → Dec (e ≤ e')) ∋ it
+
+  _≤ᵉ?_ : ∀ e e' → Dec (e ≤ᵉ e')
+  e ≤ᵉ? e' = firstSlot e ≤? firstSlot e'
```

Comment:
Yes, I thought so too... but then I can't get [this line](https://github.com/input-output-hk/formal-ledger-specifications/blob/07184c4bf41ecc896338c7836315fcb182de0575/src/Ledger/Ratify.lagda#L256) to type check (with `≤?_` in place of `≤ᵉ?_`). 🤔

Reply:
But did you provide the instances instead? Agda has to now somehow that leq for epochs is `≤ᵉ` and its decidability proof.

Reply:
Yes, I tried that... but maybe didn't do it correctly.  I'll try again!

Reply:
@omelkonian The problem is that we don't have a type class for decidable leq for Epoch since leq for Epoch is not a partial order (it's not antisymmetric)... just a preorder.  I suppose I could have made a decidable preorder type class, but instead I defined decidability of leq for Epoch in terms of decidability of leq for Slot.

Reply:
Actually that's wrong, I just didn't catch it in my first review: there is no point in requiring a stronger constraint of `HasPartialOrder` for the decidable leq, since the laws are not used in any way. So just demote the constraint to `HasPreorder` and you should be able to make the suggested change.

(also see my previous comment about how the typeclass itself is redundant, but you can keep it for now)